scripts/landcover.py

```diff
@@ -0,0 +1,202 @@
+    """
+    # compute_stats(wd, worldcover, mgrs)
+    sample(wd)
+
+
+def compute_stats(wd, worldcover, mgrs):
+    """
+    Compute statistics of Worldcover data over MGRS tiles.
+    """
+    result = []
+    with rasterio.open(worldcover) as cover:
+        with fiona.open(mgrs, "r") as tiles:
+            assert cover.crs.to_epsg() == tiles.crs.to_epsg()
+            for tile in tiles:
+                print(tile.properties.get("Name"))
+                # Split polygons in parts, mgrs tiles at the dateline
+                # are split into two parts in the mgrs source file.
+                parts = shape(tile["geometry"]).geoms
+                pixels = []
+                for polygon in parts:
+                    pixels.append(
+                        cover.read(
+                            1, window=from_bounds(*polygon.bounds, cover.transform)
+                        ).ravel()
+                    )
+                pixels = numpy.hstack(pixels)
+
+                pixels = pixels[pixels != NODATA]
+
+                if not pixels.size:
+                    continue
+                elif numpy.all(pixels == WATER):
+                    continue
+                else:
+                    props = {}
+                    for key, classname in CLASSES.items():
+                        props[str(classname)] = int(numpy.sum(pixels == key))
+                    props["name"] = tile.properties.get("Name")
+                    props["count"] = int(len(numpy.unique(pixels)))
+
+                    result.append(
+                        {
+                            "geometry": dict(tile["geometry"]),
+                            "properties": props,
+                        }
+                    )
+
+    with fiona.open(
+        Path(wd, "mgrs_stats.fgb"),
+        "w",
+        driver="FlatGeobuf",
+        crs=WGS84,
+        schema=SCHEMA,
+    ) as colxn:
+        colxn.writerecords(result)
+
+
+def split_highest(data, column, size, pool=1000, seed=RANDOM_SEED):
+    """
+    Split hightest values of a cloumn from a dataframe.
+    """
+    data.sort_values(column, ascending=False, inplace=True)
+    return data[:pool].sample(size, random_state=seed)
+
+
+def percentages(data):
+    """
+    Normalize all numerical columns to percentages
+    """
+    data_num = data.select_dtypes(include="number")
+    data_norm = data_num.div(data_num.sum(axis=1), axis=0)
+    data[data_norm.columns] = data_norm
+
+    return data
+
+
+def sample(wd):
+    """
+    Sample the mgrs tiles based on landcover statistics.
+
+    Target: ~1000 tiles
+    Set very small counts to zero. Exclude high latitudes.
+    200 samples from the 2000 most diverse
+    50 samples from the 1000 highest for all other categories except water
+    100 samples from all tiles with water between 30% an 70% (making sure we
+    capture some, but exclude only purely water so we catch coasts)
+    """
+    data = geopandas.read_file(Path(wd, "mgrs_stats.fgb"))
```

Comment:
Could you share this `mgrs_stats.fgb` file please? I got a permission error trying to access `s3://esa-worldcover/v200/2021/map` for some reason, and downloading from https://worldcover2021.esa.int/downloader is taking a long time!

Reply:
Strange `s3 sync s3://esa-worldcover/v200/2021/map $wd/esa-worldcover-v200-2021-map --no-sign-request` works perfectly for me. But yes, its quite a lot of data (120GB) so no need for everyone to reproduce the stats file! Attaching below.